File: src/util/rng.cpp
/// @file rng.cpp

#include "util/rng.h"

namespace coverwise {
namespace util {

namespace {

/// @brief 32-bit left rotation.
inline uint32_t Rotl(uint32_t x, int k) { return (x << k) | (x >> (32 - k)); }

}  // namespace

void Rng::Seed(uint64_t seed) {
  // SplitMix32: expand a single 32-bit seed into the four 32-bit state words.
  // All arithmetic is performed modulo 2^32 (native uint32_t wraparound),
  // mirroring JavaScript's Math.imul / `>>> 0` semantics.
  uint32_t state = static_cast<uint32_t>(seed);
  uint32_t words[4];
  for (int i = 0; i < 4; ++i) {
    state += 0x9e3779b9u;
    uint32_t z = state;
    z = (z ^ (z >> 16)) * 0x85ebca6bu;
    z = (z ^ (z >> 13)) * 0xc2b2ae35u;
    words[i] = z ^ (z >> 16);
  }
  s0_ = words[0];
  s1_ = words[1];
  s2_ = words[2];
  s3_ = words[3];
  // Ensure state is never all-zero.
  if ((s0_ | s1_ | s2_ | s3_) == 0) {
    s0_ = 1;
  }
}

uint32_t Rng::Next() {
  const uint32_t result = Rotl(s1_ * 5u, 7) * 9u;
  const uint32_t t = s1_ << 9;

  s2_ ^= s0_;
  s3_ ^= s1_;
  s1_ ^= s2_;
  s0_ ^= s3_;

  s2_ ^= t;
  s3_ = Rotl(s3_, 11);

  return result;
}

uint32_t Rng::NextUint32(uint32_t max) {
  if (max == 0) return 0;
  // Mirrors uniformInt(0, max - 1). When max == 1 the range is a single value,
  // so the result is 0 without consuming a draw (matching the TS short-circuit
  // `min >= max` in uniformInt). This keeps the integer stream identical.
  if (max == 1) return 0;
  // range == max.
  // Rejection sampling to avoid modulo bias.
  // threshold == (2^32 - range) % range == (-range) % range.
  const uint32_t range = max;
  const uint32_t threshold = (0u - range) % range;
  uint32_t r;
  do {
    r = Next();
  } while (r < threshold);
  return r % range;
}

}  // namespace util
}  // namespace coverwise

File: src/ts/util/rng.ts
/// @file rng.ts
/// @brief Seeded PRNG for deterministic generation using xoshiro128**.

/// SplitMix32: used to initialize xoshiro128** state from a single seed.
function splitmix32(seed: number): () => number {
  let state = seed >>> 0;
  return (): number => {
    state = (state + 0x9e3779b9) >>> 0;
    let z = state;
    z = Math.imul(z ^ (z >>> 16), 0x85ebca6b) >>> 0;
    z = Math.imul(z ^ (z >>> 13), 0xc2b2ae35) >>> 0;
    return (z ^ (z >>> 16)) >>> 0;
  };
}

/// Seeded PRNG using xoshiro128** algorithm.
///
/// Provides deterministic random number generation: same seed always produces
/// the same sequence of outputs.
export class Rng {
  private s0: number;
  private s1: number;
  private s2: number;
  private s3: number;

  /// Construct a PRNG with the given seed.
  constructor(seed: number = 0) {
    const init = splitmix32(seed);
    this.s0 = init();
    this.s1 = init();
    this.s2 = init();
    this.s3 = init();
    // Ensure state is never all-zero.
    if ((this.s0 | this.s1 | this.s2 | this.s3) === 0) {
      this.s0 = 1;
    }
  }

  /// Generate the next raw 32-bit value using xoshiro128**.
  private next(): number {
    const result = Math.imul(rotl(Math.imul(this.s1, 5) >>> 0, 7), 9) >>> 0;
    const t = (this.s1 << 9) >>> 0;

    this.s2 ^= this.s0;
    this.s3 ^= this.s1;
    this.s1 ^= this.s2;
    this.s0 ^= this.s3;

    this.s2 ^= t;
    this.s3 = rotl(this.s3, 11);

    // Keep all state as unsigned 32-bit.
    this.s0 >>>= 0;
    this.s1 >>>= 0;
    this.s2 >>>= 0;
    this.s3 >>>= 0;

    return result;
  }

  /// Generate a random integer in [min, max] (inclusive).
  uniformInt(min: number, max: number): number {
    if (min >= max) {
      return min;
    }
    const range = (max - min + 1) >>> 0;
    // Use rejection sampling to avoid modulo bias.
    // Standard debiasing: threshold = (2^32 - range) % range = (-range) % range.
    // Values below threshold map unevenly; reject them.
    const threshold = ((-range >>> 0) % range) >>> 0;
    let r: number;
    do {
      r = this.next();
    } while (r < threshold);
    return min + (r % range);
  }

  /// Generate a random uint32 in [0, max) (exclusive upper bound).
  nextUint32(max: number): number {
    if (max <= 0) {
      return 0;
    }
    return this.uniformInt(0, max - 1);
  }

  /// Select a random index weighted by the given weights array.
  ///
  /// @param weights Array of non-negative weights. At least one must be positive.
  /// @returns The selected index.
  weightedRandomIndex(weights: readonly number[]): number {
    let totalWeight = 0;
    for (let i = 0; i < weights.length; i++) {
      totalWeight += weights[i];
    }
    if (totalWeight <= 0) {
      return this.nextUint32(weights.length);
    }
    // Generate a random value in [0, totalWeight).
    // Use floating-point multiplication for uniform distribution across the range.
    const threshold = ((this.next() >>> 0) / 0x100000000) * totalWeight;
    let cumulative = 0;
    for (let i = 0; i < weights.length; i++) {
      cumulative += weights[i];
      if (threshold < cumulative) {
        return i;
      }
    }
    // Fallback for floating-point edge cases.
    return weights.length - 1;
  }

  /// Reseed the generator.
  seed(seed: number): void {
    const init = splitmix32(seed);
    this.s0 = init();
    this.s1 = init();
    this.s2 = init();
    this.s3 = init();
    if ((this.s0 | this.s1 | this.s2 | this.s3) === 0) {
      this.s0 = 1;
    }
  }
}

/// 32-bit left rotation.
function rotl(x: number, k: number): number {
  return ((x << k) | (x >>> (32 - k))) >>> 0;
}
